gitk: windows: make "Auto-select" copy to clipboard

avih opened the pull request:
The config option "Auto-select SHA1 (length)" (widget-text-)selects the first $autosellen chars of the commit hash. However, selection on its own is arguably not very useful, and also not necessarily visible when not focused, e.g. with Windows native theme (the default, but is visible on Windows with other themes, like "clam").

The real goal here to copy the text to the [selection?] clipboard, which works on X11 where "selection" is synonymous with "clipboard selection", but not necessariy elsewhere, for instance on Windows (where selecting the text doesn't copy it), and maybe elsewhere too.

This commit adds explicit copy to clipboard on Windows, which happens whenever the user clicks a commit-graph entry, and like X11, affected by the configured length and enable/disable checkbox.

Possible future modifications:
- Fix other platforms if needed, e.g. maybe non-X11 macOS Tk, Wayland.
- Rename Auto-select to Auto-copy (at least at the config dialog).
- Rename SHA1 to "commit hash" (I think git plans a move to SHA256).
- Ensure "foo bar" Auto-select length is rejected at the dialog (probably also at other numeric input fields).

avih commented:
Two unrelated notes.

Thanks for taking maintainership of gitk. IMHO it's an immensly useful tool, and it didn't see many changes in recent years (though arguably it's very useful already, and off the top of my head I don't have a wishlist).

Just a heads-up that the downstream git-for-windows gitk has few modifications, as far as I can tell mainly related to PATH handling (`which`, `exec`, `open`) and icon loading. The upstream gitk in this repo does seem to work for me after very little testing using mingw-w64 tcl/tk 8.6.11-ish (my own build), including icons, and I don't know at which cases the different path handling kicks in (I'm not a tcl guru, I only barely just manage to read existing code).

I'm not involved with the git-for-windows project and hopefully they do intend to upstream their modifications, but just in case you weren't aware, their file is here - https://github.com/git-for-windows/git/blob/main/gitk-git/gitk .

CC @dscho .

avih commented:
Just force-pushed: added `Signed-off-by:...` at the commit message.

j6t commented:
Clipboard data is precious, at times even *very* precious. I personally have never wished that the OID is copied to the clipboard automatically, *because it destroys my clipboard*. It totally goes against usual UI guidelines that the clipboard is overwritten by such an unrelated operation as one to select a different entry in a list, which is basically without the user's consent.

Fortunately, this behavior can be switched off (thank you for the reminder). IMO, the option should not even exist.

BTW, that the selection is copied to the clipboard on X11 is only half true. It overwrites that clipboard entry that can be pasted with a middle-mouse button click. But it does not overwrite that clipboard entry that applications have agreed on is used with <kbd>Ctrl</kbd>+<kbd>C</kbd> and <kbd>Ctrl</kbd>+<kbd>V</kbd>. *If* we extend this option to overwrite the clipboard, should we not also address this on other platforms (X11)?

If this feature is pursued further, it must become an opt-in with a new option. We cannot just let the existing option take possession of the clipboard without warning the user. My preference would then be: Automatically select: yes, overwrite the clipboard: no way.

avih commented:
> BTW, that the selection is copied to the clipboard on X11 is only half true. It overwrites that clipboard entry that can be pasted with a middle-mouse button click.

Yes, I know what selection is in X11/Xorg.

> We cannot just let the existing option take possession of the clipboard without warning the user.

I can't speak about the future, but it has been copying to the selection clipboard for nearly exactly two decades now. Specifically, from the very first commit which added the SHA1 widget - https://github.com/git/git/commit/98f350e50124567f90691f6142e1c048c2b4600c although it did become configurable about 17 years ago, though still enabled by defaults - https://github.com/git/git/commit/95293b58eb2ac31f93266f9e86ea17c0999d75c9 .

In general, I do agree that it's not nice to overwrite the user's whatever clipboard as soon as you start a program, and then also affectively whenever you touch it to do anything.

But specifically in this case, I do find it useful.

Did it bother you before? Did you disable it?

> If this feature is pursued further, it must become an opt-in with a new option.

No particular preference personally about being opt in or opt out, because I know how to enable it if it's opt-in, and I would.

However, are you going to disable it one day for people who have been using it for decades, expecting them to re-enable it if then want it?

Sounds like "yes", judging by "with a new option"?

I don't think it a bad thing, but I think it could be a bit annoying, but hopefully not hard to re-enable for those who prefer it.

Anyway, regardless, it would be nice to have this feature on additional platforms where possible, inluding the auto-copy whenever a new commit is selected, if they so prefer.

avih commented:
> _If_ we extend this option to overwrite the clipboard, should we not also address this on other platforms (X11)?

No opinion. I think the main thing is that if you need it, then it gets copied to someplace which can then be used conveniently. Personally I don't have an issue with using the selection clipboard with X11.

I think it's generally useful on X11 that there's more than one clipboard, but also relatively (entirely?) unique among desktop environments, at least without additional "clipboard managers" which can maintain clipboard history etc.

As far as I know, on Windows, macOS, and Wayland there's only one clipboard.

But if on some platforms there's more than one, then it probably wouldn't be a bad idea to make the target clipboard configurable.

j6t commented:
> Did it bother you before? Did you disable it?

On X11, it didn't bother me particularly, because it only overwrites the "selection", but not the clipboard. The reason is simply that I know that the "selection" is extremely volatile and I don't use it for precious data. But it is useful every now and then.

> However, are you going to disable it one day for people who have been using it for decades, expecting them to re-enable it if then want it?
> 
> Sounds like "yes", judging by "with a new option"?

It is not necessary to disable the "selection" feature in its current form. It does not overwrite the clipboard on X11 nor on Windows (I don't know of MacOS).

avih commented:
I see.

So you're OK-ish with overwriting selection on X11 by default because it's not primary, but don't like overwriting the (probably only/main) clipboard on other platforms by deafult?

If yes, then it sounds reasonable to me, and I didn't realize your main objection was depending on whether it overwrites the primary clipboard.

How do you want to proceed?

I can test on windows, but I have near zero experience with Tk or tcl. This 3-lines patch took me way longer to write than I'd like to admit.

As far as I'm concerned, I would have preferred that you add the new config to the code and dialog, and then it doesn't matter who writes the 2-lines windows code. I don't really care about the credit.

I don't have macOS to test any longer, but I believe Tk can be either native or X11, where the X11 variant probably copies the selection (possibly later into the primary macOS clipboard, which IIRC is a checkbox at the X11-for-macOS config menu), while the native variant is probably the most common variant, and it probably copies nothing.

I have no idea if there's even a native Wayland Tk and/or how common it is. I'd imagine that the common Tk is X11 via XWayland. I don't know how X11 selection is integrated with Wayland.

I don't know if there are other platforms which matter beyond those 4 (x11/win/mac/wayland).

j6t commented:
My preference would be:
 - Keep the current option that tells whether part of the commit id in the edit box is selected or not. That this overwrites the "selection" on X11 can be left as a compatibility feature.
 - Add another option (ideally dependent on that the former one is enables) that tells whether the (primary) clipboard is also overwritten at the same time. This should happen on all platforms. The option must be off by default.
 - Additionally, make a keyboard binding that copies the selected commit ID to the clipboard. This cannot be <kbd>Ctrl</kbd>+<kbd>C</kbd>, because we need that to stay the usual copy operation in the focused widget.

avih commented:
How does that look?

Note that in 3 places the text now exceeds 80 columns, but I wasn't sure if I should ensure fit, because at least in one case it would require touching quite a few following lines as well.

Let me know if you want me to ensure fit.

avih commented:
> * ideally dependent on that the former one is enables

Why is that? Is it not possible for someone to want to copy to the primary clipboard without also updating the selection clipboard?

It think it also simplifies the code to make it independent (didn't try to make it dependent) .

j6t commented:
> > * ideally dependent on that the former one is enables
> 
> Why is that? Is it not possible for someone to want to copy to the primary clipboard without also updating the selection clipboard?

The first option is to *select* something. It is just a side-effect on one platform (X11) that the selected text is also copied. 

The second option then copies the selected text to the clipboard. This makes only sense if something is selected. If the first operation is off (nothing gets selected), this operation makes no sense.

avih commented:
> The first option is to _select_ something. It is just a side-effect on one platform (X11) that the selected text is also copied.

I don't think so. I think selecting without the side effect of copying would be largely meaningless here. The widget is not even focused when one clicks a commit at the graph. I think the sole goal IS to copy the selection. I think it only uses "select" at the name because in X11 it's synonymous with selection-clipboard, and maybe also because whoever added it didn't care too much about the name.

> The second option then copies the selected text to the clipboard. This makes only sense if something is selected. If the first operation is off (nothing gets selected), this operation makes no sense.

Why? It copyes the SHA1 to the clipboard. How is selection relevant?

Do you find the behavior which the patch adds lacking or inconsistent or misleading somehow?

j6t commented:
> > The first option is to _select_ something. It is just a side-effect on one platform (X11) that the selected text is also copied.
> 
> I don't think so. I think selecting without the side effect of copying would be largely meaningless here. The widget is not even focused when one clicks a commit at the graph.

Selecting part of the text in the edit box is not meaningless if only part of the text should be copied, i.e., not the entire SHA1. It also serves as a visual cue for the third point in my list above:

> * Additionally, make a keyboard binding that copies the selected commit ID to the clipboard.

Of course, if the whole SHA1 is copied all the time, seeing a selection is less important, but then we also don't need an edit box to choose a length.

I haven't looked at the commits, yet. It will take some time; please bear with me.

avih commented:
> Selecting part of the text in the edit box is not meaningless if only part of the text should be copied, i.e., not the entire SHA1. It also serves as a visual cue for the third point in my list above:

Correct, it can also serve as a visual cue to remind the length that you configured at the prefs dialog.

But I still think the whole point here is exclussively copying (to selection clipboard). I think selecting it at the widget is only means to an end, and it being a visual cue is only side effect of the implementation details, albeit maybe not a totally useless one.

> > * Additionally, make a keyboard binding that copies the selected commit ID to the clipboard.
> 
> Of course, if the whole SHA1 is copied all the time, seeing a selection is less important, but then we also don't need an edit box to choose a length.

Maybe I'm misunderstanding something.

Let's assume for a second that the auto-select feature should have been named "auto-copy-[short-]SHA1-to-the-selection-clipboard", where the length can be configured, but that selecting the text at the widget is just implementation details (which is exactly what I think it is, but do bear with me and assume I'm right).

Is the keybind request related to adding auto-copy which does the same thing as auto-select, except to the primary clipboard instead of the selection clipboard?

If you insist that it's tightly related, then I'd consider adding it, but because I see it as a (useful) independent feature, but which I don't want to test my tcl skills with, I prefer to not add it myself.

> I haven't looked at the commits, yet. It will take some time; please bear with me.

Sure. Meanwhile, here's how it looks. Both checkboxes are independent, and the length applies equally to both.

One (selects and) copies the configured length SHA1 to the selection clipboard, the other copies the configured length SHA1 to the primary clipboard:

![gitkcfg](https://github.com/user-attachments/assets/26b86b1f-569a-4fbc-b280-2cc4df50d15a)

I don't feel strong at all about the text at the labels, but I did try to make it reasonably useful.

avih commented:
> > > The first option is to _select_ something. It is just a side-effect on one platform (X11) that the selected text is also copied.
> > 
> > 
> > I don't think so. I think selecting without the side effect of copying would be largely meaningless here. The widget is not even focused when one clicks a commit at the graph.
> 
> Selecting part of the text in the edit box is not meaningless if only part of the text should be copied, i.e., not the entire SHA1. It also serves as a visual cue for the third point in my list above:

Think about it this way:

What useful thing does Auto-select do on Windows or macOS - where there's no selection clipboard?

And my answer is: nothing. Do you disagree?

Because the only useful effect it has is that the selection clipboard is populated with the [short] SHA1, so that you can paste it elsewhere and use it.

Without this effect it's useless.

Therefore the only real goal of "select" is "populate the selection clipboard".

avih commented:
> I agree that this option does nothing useful _currently_ on Windows and macOS.

The whole point of this side discussion was to hopefully reach an agreement (which I think we've done) that the premise of Auto-select is not to select (with a nice side effect on X11 of populating the selection clipboard), but rather to populate the selection clipboard, with (a marginally useful) side effect of it being visibly selected due to implementation details.

And it's not even always visibly selected. On windows native Tk theme, which is the default, the selection is not visible because it's only visible in focused widgets, which the SHA1 widget is not when clicking a commit at the graph. So Windows users would only see the selection if they change the tk theme, for instance to clam, which I bet 99% if not more of Windows gitk users never do (because it's not obvious at all how, it happens to work by creating `~/.Xdefaults` X11 resource file and populating it accordingly, which Tk happens to load also on Windows. Also it doesn't work with tcl/tk 9, though neither does gitk currently). But overall, no issue, because "visibly selected" is only marginally useful.

I don't know whether it's visible in macOS.

Bottom line, the goal is to populate the selection clipboard when working in X11 env.

Next.

> * They can have enabled a second option so that the highlighted part is copied to the clipboard automatically every time a commit is selected.
> 
> * Or they can not have  the second option enabled, but use a keyboard shortcut similar to Ctrl+C (but not exactly that) instead to copy the highlighted part of the SHA1 to the clipboard.
> 
> * Or they can do nothing (and the clipboard is not overwritten).
>
> Now it is obvious that this second option depends on the existing

I agree with all of those, but it's still not obvious to me why they're dependent.

Here's my alternative suggested functionality:
- There's a configurable SHA1 length which is used with the various copying methods. This already exists today.
- Auto-select populates the selection clipboard with it - just like now, and only works with X11 (with some visibility side efffect on some platforms).
- Auto-copy populates the primary clibpard with it, and works in all platforms, including X11 (where it targets a different clipboard than Auto-select), and is independent of Auto-select.
- Some KB binding populates the TBD clipboard with it, which is independent of both Auto-select and Auto-copy, and doesn't require nor depend on any checkboxe.

The first two items already work today in gitk master,

This PR currently adds the third one - Auto-copy, with the semantics/behavior as described above (please at least glance at the PR code. The commit is two lines for the behavior, and two lines for the widget, and the 2nd commit only trivially touches the Auto-select label text).

The additional KB binding would definitely be nice to have, but I feel it's beyond my Tcl/Tk skills, so I'd prefer to not add it myself.

Here's why I think it's better:
- Options dependency adds mental strain for the user, and more complex code and widget for the maintainer.
- It allow using Auto-copy on X11 without Auto-select.
- If you only care about Auto-copy, then you don't have to wonder why it doesn't work when Auto-select is not checked (yes, you can disable the widget when Auto-select is disabled, but why?).
- If you only care about Auto-copy, then you only need to care about one checkbox rather than too.
- You don't lose anything compared to making them dependent.

Do you not think this is overall more complete and easier to use than making them dependent?

avih commented:
And, If we agree on independence, then we can consider retouching the layout a bit, to make it more obvious that the length doesn't apply only to Auto-select, and that Auto-select does actually perform some copying, for instance like so (I don't care much about the wording, do feel free to suggest something better):

![gitk-copy-layout](https://github.com/user-attachments/assets/51289848-30ba-4e9a-89df-69d195d3c51a)

j6t commented:
OK, let's do it like you suggest. My suggested wording and order of controls is:

☑ Copy commit ID to clipboard
☑ Copy commit ID to X11 selection
Length of commit ID to copy  _______

Bonus points if the "X11 selection" option is only available on X11, because it is a hell lot of user-unfriendly nerd lingo that should not be visible if it isn't applicable.

avih commented:
> My suggested wording and order of controls is:

Thanks. LGTM. Do notice that elsewhere in gitk it still uses "SHA1". I think that for consistency they should all be either SHA1 or "Commit ID" (or maybe "SHA" ?)

Do you want to address it? if so, how? I don't mind adding a commit which changes it elsewhere too.

> Bonus points if the "X11 selection" option is only available on X11, because it is a hell lot of user-unfriendly nerd lingo that should not be visible if it isn't applicable.

Agreed. I'll do that. Do you want it removed from the dialogue in such case? or grayed out? (I think remove is fine, but I don't mind which)

Here are my suggested commits:
1. If we want to change SHA1, then change all SHA1 to "Commit ID" (or "SHA", let me know).
2. Change the Auto-select text and layout as suggested above (without auto-copy), and enable it (checkbox and length) only on X11
3. Add the Auto-copy commit, and ensure that the length is now visible unconditionally (while Auto-select remains conditional).

Let me know whether that's OK or if you prefer something else.

avih commented:
Also, how to we want to detect x11? that we have env DISPLAY?

avih commented:
By the way, I checked it out, and it turns out that on Wayland there's both primary and selection clipboards.

I don't know whether there's native tk for Wayland, but if there is, then the X11 detection would miss that (at least DISPLAY is not set in Wayland unconditionally. I believe it's set only when XWayland is available - at which case we can't know by DISPLAY alone whether Tk is X11 or Wayland).

The question we're really trying to answer is "does selecting text at the widget populate the selection clipboard?"

And I don't know how to answer that.

May I suggest that we leave that out for now (or forever), and simply keep both elements enabled unconditionally?

j6t commented:
> 1. If we want to change SHA1 elsewhere, then change all SHA1 to "Commit ID" (or "SHA", let me know).

Good. It is desirable to change "SHA1" in the UI to "Commit ID" in a preparatory commit first. This is an uncontroversal change (it must also change the instance in the Options dialog).

>   2. Change the Auto-select text and layout as suggested above (without auto-copy), and enable it (checkbox and length) only on X11
>
>   3. Add Auto-copy, and ensure that the length is now enabled unconditionally (while Auto-select remains conditional).

Ok, but move the removal of the X11-control on non-X11 into a step 4.

About Wayland:

> May I suggest that we leave that out for now (or forever), and simply keep both elements enabled unconditionally?

Please check just the DISPLAY variable for now. We can ask on the mailing list whether users know about the secondary clipboard on Wayland, how it behaves, and whether they use or need it, and then think about a solution if that is the case.

avih commented:
> Ok, but move the removal of the X11-control on non-X11 into a step 4.

Sounds good. Do you want it removed or disabled (grayed out)? Both are trivial, I think I'd pick "remove" because it doesn't change at the lifetime of a gitk instance (or anytime on the platform), but again, I'm fine with either.

avih commented:
All done as requested, as far as I can tell. Few notes:

1. I did the name change as requested:

> It is desirable to change "SHA1" in the UI to "Commit ID" in a preparatory commit first

and I'm not against it, but I think it ended up a bit awkward, mainly due to the capitalization, but also because I think it's a bit long-ish.

Maybe reconsider changing it to "commit ID" or "Commit-ID" or "SHA-ID" or "SHA ID"?

Personally I'd go with "SHA ID", because I think SHA will remain in git for at least few decades to come, it's recognizable and stands out better IMHO, and it's also shorter in UI text, which I think reduces mental strain.

2. At the autocopy commit I had to add `autocopy` as global in few places, and placed it adjacent to `autosel` for readability (prior to, to match the UI order we preferred), and also made some adjustments to keep lines length of 80 columns.

In one place it fit without issues, in one I moved `tabstop` to the previous line, and in one place I didn't change other lines and the line exceeds 80 columns. Alternatives are placing `autocopy` at the end of the list - few lines below `autosel`, or touching all the following lines and moving the last var in each line to the next line.

Let me know if you want to change this.

3. At the commit message of "show autosel pref only in X11", I added "This feature was requested by Johannes Sixt", because I wouldn't have added it myself, but I'm OK with it when it includes this info. Alternative might me to drop it from this PR, and you can add it yourself after you merge it.

Let me know if you want me to remove this commit.

j6t commented:
> 1. I did the name change as requested:
> 
> > It is desirable to change "SHA1" in the UI to "Commit ID" in a preparatory commit first
> 
> and I'm not against it, but I think it ended up a bit awkward, mainly due to the capitalization, but also because I think it's a bit long-ish.
> 
> Maybe reconsider changing it to "commit ID" or "Commit-ID" or "SHA-ID" or "SHA ID"?

My intent was to use "commit ID" or "Commit ID" as needed in each instance. That name, BTW, is the official term for a commit's hash value. So, it should not be "SHA" or related.

> 2. At the autocopy commit I had to add `autocopy` as global in few places, and placed it adjacent to `autosel` for readability (prior to, to match the UI order we preferred), and also made some adjustments to keep lines length of 80 columns.

I do not insist in the 80 column limit at all as long as the result is within reason. So, I would have preferred that unrelated lines would not have been changed when possible. (But you could not have known that.)

I appreciate that you put related variables together.

> 3. At the commit message of "show autosel pref only in X11", I added "This feature was requested by Johannes Sixt", because I wouldn't have added it myself, but I'm OK with it when it includes this info. Alternative might be to drop it from this PR, and you can add it as your own after you merge it.

OK. It's a notice that is irrelevant to later readers of the patch. I'll decide later.

avih commented:
> My intent was to use "commit ID" or "Commit ID" as needed in each instance.

I see. I'll do that. That would look better indeed. Thanks.

> I do not insist in the 80 column limit at all as long as the result is within reason. So, I would have preferred that unrelated lines would not have been changed when possible.

So, the patch already has one line which exceeds 80 cols because I didn't touch other lines, and there's only one place where I changed another line - I move one var to the previous line.

Sounds like you prefer to move that var back to the same line, and we'll have two lines which exceed 80?

j6t commented:
Regarding the commits:

1. In "gitk: UI text: change "SHA1 ID" to "Commit ID":
   - The term "commit ID" is not less specific (it's actually more specific), but the official name.
   - As said earlier here, we should lower-case some of the instances of the new word "Commit" (but not "ID").
2. In "gitk: prefs dialog: refine Auto-select UI":
   - The commit message uses a mix of "Commit ID", "commit-ID"
   - The reason to move the length edit box to a new line has changed: It's now because we are going to introduce another option that will use the value.
   - I like that you moved "Hide remote refs" up.
3. In "gitk: support auto-copy SHA to primary clipboard":
   - It would have been nice if you would have broken the `checkbutton $page.autocopy` line in the same manner as the `checkbutton $page.autoselect`, but no big deal.
4. In "gitk: prefs: show Copy-to-X11-selection only in X11":
   - The "requested by" sentence feels awkward in the commit message.

I like the result. Do you want to apply these touch-ups?

avih commented:
> I like the result. Do you want to apply these touch-ups?

Thanks. Sure. Let me just make sure I understand what you want or don't want to change:
- Chage "less specific" to "the official name" or some such at the reason for changing SHA1.
- Use "Commit ID" consistently in commit messages, capitalized only at the begining of a sentence (which there are none, so all would be "commit ID").

> The reason to move the length edit box to a new line has changed: It's now because we are going to introduce another option that will use the value.

I'm not sure whether this is a request for a change, or confirming that the reason at the commit message is OK. It does mention that it's because we'll be adding another option to use it, but also that there _already is_ another place where it's used.

Here's what it says now:
```
Additionally, move the length input widget to a new line, because:
- This length applies to both Auto-select and "Copy commit reference"
  menu item, so it's not exclusive to to the selection length.
- The next commit will add support for primary clipboard as well,
  where this length will also be used.
```

Do you want to remove the first reason from the text, and keep only the "the next commit will add..." part?

> It would have been nice if you would have broken the checkbutton $page.autocopy line in the same manner as the checkbutton $page.autoselect, but no big deal.

> Actually, the other way round. All other options (and new ones from a different topic) prefer to have the label on the line with the `checkbutton` command and the `-textvariable` at the beginning of the next line.

It doesn't fit 80 columns if we keep the text at the first line. That's why I wrapped it.

Do you prefer to move it back and have it exceed 80 columns?

Or, alternatively, apply your original suggestion and wrap the other line the same way?

Additionally, I asked specifically few times already, but you didn't say anything concrete about the existing line which already exceed 80 columns, and the other place where I moved `tabstop` to the previous line to avoid exceeding 80 columns.

I'm absolutely fine doing whatever you prefer, but I'm not convinced I understand it.

Should I assume you're OK with both and prefer to not touch it?

> * The "requested by" sentence feels awkward in the commit message.

Is this a request to keep the commit, but also remove that line from the commit message?

If yes, then as I said, I'm uncomfortable having it as my own commit, because I wouldn't have added it myself. I don't like this commit.

If you really don't want this line, then I can change the commit author to be you (fake it to save you the trouble of writing your own commit), and you'll add your signoff before you merge.

But really, it's a trivial commit, code-wise, which you want and I don't. Wouldn't it be simplest to drop it from this PR and you'll add it as your own later?

I'm not against you adding it later. I just don't want it to appear as my own commit.

j6t commented:
> ```
> Additionally, move the length input widget to a new line, because:
> - This length applies to both Auto-select and "Copy commit reference"
>   menu item, so it's not exclusive to to the selection length.
> - The next commit will add support for primary clipboard as well,
>   where this length will also be used.
> ```
> 
> Do you want to remove the first reason from the text, and keep only the "the next commit will add..." part?

Yes, remove the first reason, because at this point in the series it is incorrect. You could mention in the next commit that introduces the option that it uses the same length value as the existing option.

> It doesn't fit 80 columns if we keep the text at the first line. That's why I wrapped it.
> 
> Do you prefer to move it back and have it exceed 80 columns?

Yes, please move it back. The 80 columns limit is ... undesirable. My editor, and probably yours, too, can show much longer lines without wrapping. (But please don't be excessive.)

> Additionally, I asked specifically few times already, but you didn't say anything concrete about the existing line which already exceed 80 columns, and the other place where I moved `tabstop` to the previous line to avoid exceeding 80 columns.

Please do not move  `tabstop` up. The 80 column limit does not exist in this code base. (But please don't take this as excuse to be excessive.)

> > * The "requested by" sentence feels awkward in the commit message.
> 
> Is this a request to keep the commit, but also remove that line from the commit message?

Yes, exactly that.

> If yes, then as I said, I'm uncomfortable having it as my own commit, because I wouldn't have added it myself. I don't like this commit.

Fair enough. Please drop the commit.

avih commented:
Much appreciated. I'll do these changes.

avih commented:
> Yes, remove the first reason, because at this point in the series it is incorrect.

Wait.

Why is it incorrect? If you right click on a commit at the graph (in gitk master, before this PR), you have an option menu "Copy commit reference". This option already uses the length.

So it states one reason that it's already used in two places.

And another reason that the next commit will add yet a third place which uses the same length value.

j6t commented:
> > Yes, remove the first reason, because at this point in the series it is incorrect.
> 
> Wait.
> 
> Why is it incorrect? If you right click on a commit at the graph (in gitk master, before this PR), you have a menu item "Copy commit reference". This action already uses the length - in addition to Auto-select which also uses it.

My bad, you are right, of course. I didn't read the sentence carefully enough. The text is alright. I'm sorry.

avih commented:
OK, updated everything as requested.

Do let me know if I missed something, and thanks for the patience and understanding.

avih commented:
Force-pushed: forgot and fixed one "Commit ID" to "commit ID" at the autocopy commit.

avih commented:
```
- This length applies to both Auto-select and "Copy commit reference"
  menu item, so it's not exclusive to to the selection length.
```
The patchset above includes a typo fix "to to" -> "to", and I changed "menu" into "context meu" for further clarity.

Other than those,  the set only has the requested changes.

j6t commented:
Thank you, this looks very good. I'll give these changes a try on Windows and merge them later.

avih commented:
Thanks.

Unrelated, reminder that you might want to also glance at the git-for-windows downstream changes - all seem very recent - https://github.com/git-for-windows/git/commits/main/gitk-git/gitk

(these 5 patches from 3 weeks ago appear to be the only diff - `gitk` after their commit "[doc: switch links to https](https://github.com/git-for-windows/git/commit/d05b08cd52cfda627f1d865bdfe6040a2c9521b5)" is identical to `gitk` at ed87b13 in this repo, except missing newline at EOF in one of them)

avih commented:
> I can't reproduce the left/right arrow keys issue in gitk in this repo which one of their patches addresses. I'm using my own build of tcl/tk 8.6.11-ish, and key-left/key-right do work for me as history back/forward.

I can actually reproduce it, and as far as I can tell it's not limited to Windows.

When inside the SHA1 widget (soon to be Commit ID), the left/right arrow keys do nothing with tcl/tk 8.6.X with gitk master on windows and linux, but do move the cursor with this patch.

Not critical, because typically one doesn't need to navigate within the hash (especially not when using autoselect/copy), but generally the cursor keys are expected to work at text input widgets, so it's still nicer if the cursor can be moved.
